kvserver: disconnect rangefeeds before applying a merge

ajwerner opened the pull request:
This commit changes where we disconnect rangefeeds in the face of a merge
trigger. Prior to this commit, we'd disconnect the rangefeed after applying
the merge to the engine and, importantly, after feeding the logical ops to
the rangefeed (which happens in pre-apply). Now we'll disconnect the rangefeed
from both sides prior to applying the command and prior to sending the logical
ops.

No release note is needed because the bug that this fixes couldn't be tickled
before the 20.2 release as it was not possible to create a rangefeed on the
system config span as a client and the system never did it.

Fixes #48770.

Release note: None

Commit:
kvserver: disconnect rangefeeds before applying a merge or split

This commit changes where we disconnect rangefeeds in the face of a split or
merge trigger. Prior to this commit, we'd disconnect the rangefeed after
applying the merge to the engine and, importantly, after feeding the logical
ops to the rangefeed (which happens in pre-apply). Now we'll disconnect the
rangefeed from both sides prior to applying the command and prior to sending
the logical ops.

No release note is needed because the bug that this fixes couldn't be tickled
before the 20.2 release as it was not possible to create a rangefeed on the
system config span as a client and the system never did it.

Fixes #48770.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/49083)
<!-- Reviewable:end -->

nvb reviewed (commented):
> Reviewed 4 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49083)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/kv/kvserver/client_rangefeed_test.go, line 107 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49083#-M7JxFRTDyROuuNvA-GV:-M7JxFRU0CBqNkrfiHmc:b-sq6zsj) ([raw file](https://github.com/cockroachdb/cockroach/blob/d151edb55cf38f4d90703a1e7ebe8989367ccfe4/pkg/kv/kvserver/client_rangefeed_test.go#L107)):*
> ```Go
> }
> 
> // TestMergeOfRangeEventTableWhileRunningRangefeed ensures that it is safe
> ```

This sentence sounds off. It says "is safe" twice.
___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 649 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49083#-M7JxTSM1BgFBcq27snq:-M7JxTSM1BgFBcq27snr:bzdqvso) ([raw file](https://github.com/cockroachdb/cockroach/blob/d151edb55cf38f4d90703a1e7ebe8989367ccfe4/pkg/kv/kvserver/replica_application_state_machine.go#L649)):*
> ```Go
> 		}
> 
> 		// Shut down rangefeed processors on either side of the merge.
> ```

It seems a little unfortunate to me that we're now in this state where some of the `RangeFeedRetryError_REASON`s are delivered to the rangefeed processor before their corresponding Raft entry and some are delivered to it after. Notably, we still have `REASON_REPLICA_REMOVED` and `REASON_RANGE_SPLIT` being delivered after. 

I think `REASON_REPLICA_REMOVED` needs to stay where it is because it's driven from other places than the Raft log, but I'm wondering if we should move the `REASON_RANGE_SPLIT` disconnect here as well in the name of symmetry and reducing complexity.

Or we could move both here since the "NB" comment below indicates that disconnect is idempotent. We already detect "if this command will remove us from the range." later in this function.

What do you think?
___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 668 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49083#-M7Jyl-s9n3WIjpt1mAv:-M7Jyl-s9n3WIjpt1mAw:bxjd2mo) ([raw file](https://github.com/cockroachdb/cockroach/blob/d151edb55cf38f4d90703a1e7ebe8989367ccfe4/pkg/kv/kvserver/replica_application_state_machine.go#L668)):*
> ```Go
> 			roachpb.RangeFeedRetryError_REASON_RANGE_MERGED,
> 		)
> 		rhsRepl.disconnectRangefeedWithReason(
> ```

Consider adding `rhsRepl.raftMu.AssertHeld()` right above this line to annotate that the raft mu is, in fact, held.

Or even better, add it after `rhsRepl, err := ...` and comment about how we're already holding the raft mu because we are holding the merge lock in `cmd.splitMergeUnlock`.


<!-- Sent from Reviewable.io -->

Commit:
kvserver: fix race when checking whether rangefeeds are enabled

Drive-by from stressing the previous commit. This bug was introduced in
the commit which enabled rangefeeds for system ranges. It was careless.

Release note: None

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49083)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[pkg/kv/kvserver/client_rangefeed_test.go, line 107 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49083#-M7JxFRTDyROuuNvA-GV:-M7KCSQyC1-z9b0aL1jE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d151edb55cf38f4d90703a1e7ebe8989367ccfe4/pkg/kv/kvserver/client_rangefeed_test.go#L107)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This sentence sounds off. It says "is safe" twice.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 649 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49083#-M7JxTSM1BgFBcq27snq:-M7KCT7p4M0YkCabpHZG:bl2vsjd) ([raw file](https://github.com/cockroachdb/cockroach/blob/d151edb55cf38f4d90703a1e7ebe8989367ccfe4/pkg/kv/kvserver/replica_application_state_machine.go#L649)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

It seems a little unfortunate to me that we're now in this state where some of the `RangeFeedRetryError_REASON`s are delivered to the rangefeed processor before their corresponding Raft entry and some are delivered to it after. Notably, we still have `REASON_REPLICA_REMOVED` and `REASON_RANGE_SPLIT` being delivered after. 

I think `REASON_REPLICA_REMOVED` needs to stay where it is because it's driven from other places than the Raft log, but I'm wondering if we should move the `REASON_RANGE_SPLIT` disconnect here as well in the name of symmetry and reducing complexity.

Or we could move both here since the "NB" comment below indicates that disconnect is idempotent. We already detect "if this command will remove us from the range." later in this function.

What do you think?
</blockquote></details>

Seems fine to me. Moved the split logic up and re-titled the commit.
___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 668 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49083#-M7Jyl-s9n3WIjpt1mAv:-M7KCXwl4DQ54uPx6jW3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d151edb55cf38f4d90703a1e7ebe8989367ccfe4/pkg/kv/kvserver/replica_application_state_machine.go#L668)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Consider adding `rhsRepl.raftMu.AssertHeld()` right above this line to annotate that the raft mu is, in fact, held.

Or even better, add it after `rhsRepl, err := ...` and comment about how we're already holding the raft mu because we are holding the merge lock in `cmd.splitMergeUnlock`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ajwerner commented:
TFTR!

bors r=nvanbenschoten
